### DISCUSSION TITLE
How to play Videofile?

### DISCUSSION THREAD

**Discussion by GFA-SkyRanger** [kibook/pmms#44] [Q&A]
Hi,

is it possible to play videofile,located insithe the resourcefolder of pmms?
or somwhere else, may me an mp4 file on a website.

so that i can play logo-file for the Server w/o YT.

**Comment by kibook** [comment]
You can place video files in the `http/media/` folder inside the resource directory, and those files can then be played by simply specifying the name of the file.

For example, if you placed a file named `myvideo.mp4` inside `http/media/`, you can enter `myvideo.mp4` in the URL field of the pmms UI to play that video.

You may alternatively specify the full URL as well, which will be something like `http://[your server IP]:[your server port]/pmms/media/myvideo.mp4`, or `https://[owner]-[server ID].users.cfx.re/pmms/media/myvideo.mp4`. The latter is an HTTPS proxy provided by Cfx for all servers that authenticate with a license. `[owner]` is your name on the Cfx forums, and `[server ID]` is the ID assigned with your license (see https://keymaster.fivem.net).

Refer to the [httpmanager](https://github.com/kibook/httpmanager) resource for more info, that is what pmms is using to host files from inside the resource directory.

**Reply by GFA-SkyRanger** [reply]
Thx for the answer.
oh got it.

the dui folder is empty. but u have a link to another resource, this is not copied by cloning this resource..

**Reply by kibook** [reply]
The dui folder shouldn't be empty. It's a [submodule](https://git-scm.com/book/en/v2/Git-Tools-Submodules) to [another repository on GitHub](https://github.com/kibook/pmms-dui). If you clone the repository to download pmms, you need to specify `--recurse-submodules` to also clone all submodules in the repository. You can also do `git submodule update --init` to clone all the submodules after the fact if you didn't specify `--recurse-submodules` at the initial cloning.

**Reply by kibook** [reply]
The reason I broke out the DUI page content into a submodule is because it's both included in pmms to provide a plain HTTP DUI page (via httpmanager), _and_ it's hosted here with GitHub Pages to provide an HTTPS version of the page (https://kibook.github.io/pmms-dui).

Browsers don't really like HTTPS pages embedding HTTP resources (because it breaks the security provided by HTTPS) and YouTube doesn't like HTTP pages embedding HTTPS YouTube videos (it will block requests thinking it's some kind of bot), so to avoid those kinds of issues, which DUI page is used (the local HTTP or GitHub Pages HTTPS one) is determined by the protocol of the URL you specify. If you play a URL that starts with `http://`, itwill use the local HTTP DUI page, and if you play a URL starting with `https://`, it will use the remote HTTPS DUI page.

And to avoid having the same code in two different places maintained separately, I just included the GitHub pages DUI repository as a submodule inside of the main pmms repository.

**Reply by GFA-SkyRanger** [reply]
ok, understood. 
well, as i cloned the resource pmms, it wasnt kopied so i just found it and cloned it right into the dui folder now i get an format error.
which is not really helpful.

wat is meant with "format"?
Aspect-ratio, resolution, container (avi, mkv, mp4) or the codec (like h254, divx, etc...) ?

**Reply by kibook** [reply]
It's an error reported by MediaElement.js I think, and it generally refers to the container or codecs, not something like aspect ratio or resolution. You can also get a "format error" if the link isn't returning a video file at all (for example, if you specify a video file that doesn't exist and the HTTP server returns a 404 page).

What kind of video file are you trying to play?

**Reply by GFA-SkyRanger** [reply]
just a short clip i made.

`  Stream #0:0[0x1](eng): Video: h264 (High) (avc1 / 0x31637661), yuv420p(tv, bt709, progressive), 1920x1080 [SAR 1:1 DAR 16:9], 7248 kb/s, 25 fps, 25 tbr, 25k tbn (default)

  Stream #0:1[0x2](eng): Audio: aac (LC) (mp4a / 0x6134706D), 48000 Hz, stereo, fltp, 160 kb/s (default)`
Container mp4

**Reply by kibook** [reply]
Does it work if you open it in an external web browser with the full URL? Or using the full URL in pmms? Sounds similar to the issue in https://github.com/kibook/pmms/discussions/35

**Reply by GFA-SkyRanger** [reply]
aaaaah.....
ok 

You can also get a "format error" if the link isn't returning a video file at all (for example, if you specify a video file that doesn't exist and the HTTP server returns a 404 page).

this was the solution

the files doesnt have to be in the dui folder.
he just looking for his own files in the dui folder

the files still remain in media folder

got it to work!


conclution: i missed the dui-files. thats it...

thank you for helping me

next cinema-session will come.. i think, there is more potential :-)

**Reply by kibook** [reply]
No problem! Glad you got it working. Hopefully this solution will work for #35 as well.

**Reply by GFA-SkyRanger** [reply]
he got the DEMUXER ERROR...
which i get, if i try to play an file which is not h.264

it seems, it has to be h.264 to work withing an mp4 container.
just tested some files.
its just my own conclution, not sure, if it works for him.

i know video coding, and containers.. so i dont know, how much he knows about it.

**Reply by kibook** [reply]
It depends on what the embedded web browser that FiveM uses (which is a version of Chromium) and what MediaElement.js support. MPEG-4 with H.264 + AAC is usually supported because of how popular it is on the web, but technically you might be better off with a non-proprietary format like WebM with VP9 + Opus. I think the semi-official stance of the FiveM developers is they only support open formats like that, and support for proprietary ones like MPEG-4 (even if it's one of the most common in use) is not guaranteed:

https://discord.com/channels/192358910387159041/701807973793792062/729688655819702342
> 2. NUI audio/video issues. This is due to use of proprietary codecs (MP3/H.264/MP4), these are not supported in Chromium and only worked in old M73 by coincidence. Not likely to be fixed, ask Google for supporting these. MPEG LA patents are way out of our budget, and alternately we'd have to block downloading of FiveM/RedM in the US and other regions where software patents are legal. Also, some OGG files are badly converted and still have mp4 tags in them.
